Source/OCMock/OCMockObject.m

```diff
@@ -261,50 +284,75 @@ - (void)forwardInvocation:(NSInvocation *)anInvocation
     }
     @catch(NSException *e)
     {
-        [exceptions addObject:e];
+        @synchronized(exceptions)
+        {
+            [exceptions addObject:e];
+        }
         [e raise];
     }
 }
 
 - (BOOL)handleInvocation:(NSInvocation *)anInvocation
 {
-    [invocations addObject:anInvocation];
-
+    @synchronized(invocations)
+    {
+        // When the method has a char* argument we do not retain the arguments. This makes it possible
+        // to match char* args literally and with anyPointer. Not retaining the argument means that
+        // in these cases tests that use their own autorelease pools may fail unexpectedly.
+        if(![anInvocation hasCharPointerArgument])
+            [anInvocation retainArguments];
+        [invocations addObject:anInvocation];
+    }
+    
```

Comment:
Yes this change definitely caused #245.  Trying to do -[NSInvocation retainArguments] when the selector is one of the reference counting ones (i.e. -retain, -retainCount, -release, -autorelease) causes all kinds of problems.  I fixed it by adding those to the black list which I think is the right thing to do anyway.  The Objective-C runtime increasingly wants to handle those methods all by itself (or at least it really doesn't want people to be overriding them anymore), and ARC makes when they get called relatively unpredictable too.  I also copied the unit test from #245 into the project and rebased from master.